src/vanilla.ts

```diff
@@ -49,167 +48,195 @@ const snapshotCache = new WeakMap<
-          const errorOrPromise = (value as any)[PROMISE_ERROR] || value
-          Object.defineProperty(snapshot, key, {
-            get() {
-              if (PROMISE_RESULT in value) {
-                return (value as any)[PROMISE_RESULT]
-              }
-              throw errorOrPromise
-            },
-          })
-        }
-      } else if (value?.[LISTENERS]) {
-        snapshot[key] = value[SNAPSHOT]
+  const snapshot: any = Array.isArray(target)
+    ? []
+    : Object.create(Object.getPrototypeOf(target))
+  markToTrack(snapshot, true) // mark to track
+  snapshotCache.set(receiver, [version, snapshot])
+  Reflect.ownKeys(target).forEach((key) => {
+    const value = Reflect.get(target, key, receiver)
+    if (refSet.has(value)) {
+      markToTrack(value, false) // mark not to track
+      snapshot[key] = value
+    } else if (value instanceof Promise) {
+      if (PROMISE_RESULT in value) {
+        snapshot[key] = (value as any)[PROMISE_RESULT]
       } else {
-        snapshot[key] = value
+        const errorOrPromise = (value as any)[PROMISE_ERROR] || value
+        Object.defineProperty(snapshot, key, {
+          get() {
+            if (PROMISE_RESULT in value) {
+              return (value as any)[PROMISE_RESULT]
+            }
+            throw errorOrPromise
+          },
+        })
       }
-    })
-    return Object.freeze(snapshot)
-  }
+    } else if (value?.[LISTENERS]) {
+      snapshot[key] = value[SNAPSHOT]
+    } else {
+      snapshot[key] = value
+    }
+  })
+  return Object.freeze(snapshot)
+}
+
+const createProxy = <T extends object>(
+  initialObject: T,
+  handler: ProxyHandler<T>
+): readonly [T, T] => {
   const baseObject = Array.isArray(initialObject)
     ? []
     : Object.create(Object.getPrototypeOf(initialObject))
-  const handler = {
-    get(target: T, prop: string | symbol, receiver: any) {
-      if (prop === VERSION) {
-        return version
-      }
-      if (prop === LISTENERS) {
-        return listeners
-      }
-      if (prop === SNAPSHOT) {
-        return createSnapshot(target, receiver)
-      }
-      if (prop === HANDLER) {
-        return handler
-      }
-      return Reflect.get(target, prop, receiver)
-    },
-    deleteProperty(target: T, prop: string | symbol) {
-      const prevValue = Reflect.get(target, prop)
-      const childListeners = prevValue?.[LISTENERS]
-      if (childListeners) {
-        childListeners.delete(popPropListener(prop))
+  const proxyObject = new Proxy(baseObject, handler)
+  return [baseObject, proxyObject]
+}
+
+const identity = <T>(x: T) => x
+
+const buildProxyFunction = (
+  customizeObjectIs: (fn: typeof Object.is) => typeof Object.is = identity,
+  customizeCanProxy: (fn: typeof canProxy) => typeof canProxy = identity,
+  customizeCreateSnapshot: (
+    fn: typeof createSnapshot
+  ) => typeof createSnapshot = identity,
+  customizeCreateProxy: (
+    fn: typeof createProxy
+  ) => typeof createProxy = identity
+) => {
```

Comment:
Maybe change this to argument object to be less unwieldy:

```ts
const buildProxyFunction = ({
  objectIs = Object.is,
  canProxy = canProxy,
  createSnapshot  = createSnapshot,
  createProxy = createProxy, // default methods
} : {
   objectIs: typeof Object.is,
  canProxy: typeof canProxy,
  createSnapshot: typeof createSnapshot,
  createProxy: typeof createProxy, // types 
}) => {
```

Reply:
Then, you can't reuse the existing functions. One point of this style is to avoid exporting other functions (and symbols).

Reply:
I think exporting the internals would be better, since to reuse existing functions you already need to know how they work.  

Also since custom proxy is marked as unstable, developers could also understand that applies to exported internals too.

Reply:
Do you mean to export many unstable functions? Anyway, if this is just a preference in styles rather than capability, I'd go with this for this time.

Reply:
Yes. Yeah it’s a preference in style only.